### ISSUE TITLE
SSH prompts for password in Docker

### ISSUE THREAD

**Issue by prologic** [gogits/gogs#2409]
Not sure how to debug this through s6; not even sure where the typical `/var/log/auth` is?

I'm running gogs via Docker. Here is my docker-compose.yml:
```#!yml
gogs:
  image: gogs/gogs
  volume_driver: local
  volumes:
    - gogsdata:/data
  restart: always
```

When I clone a repo I just get prompted for a password:
```#!bash
$ git clone git@git.local:prologic/test.git
Cloning into 'test'...
git@git.local's password:
```

I added my public `id_rsa.pub` key contents via the Web UI -> Settings -> SSH Keys

**Comment by Unknwon** [comment]
Please check the `auth.log` of SSH server.

Docker sometimes changes the directory permission which lead to auth fail.

**Comment by prologic** [comment]
Where is this ``auth.log`` file? :) I don't mean to be silly or pretentious; but there is nothing in ``/var/log``:
```#!bash
$ docker exec -i -t f205dcab24d9 /bin/bash
bash-4.3# ls -l /var/log
total 0
bash-4.3#
```

**Comment by prologic** [comment]
Also:
```#!bash
bash-4.3# find / -type f -name 'auth.log' | wc -l
0
```

**Comment by Unknwon** [comment]
This should help: http://stackoverflow.com/questions/22526016/docker-container-sshd-logs

**Comment by prologic** [comment]
I think you've mis-read this. I'm more than happy to dig around in the running container to debug this; but I shouldn't have to! Why aren't there logs setup already in the `Dockerfile`? As it stands right now; no-one can just ``docker run ... gogs/gogs`` and expect to get a working git+ssh experience. It's broken in 0.8.10 and latest.

**Comment by Unknwon** [comment]
> As it stands right now; no-one can just docker run ... gogs/gogs and expect to get a working git+ssh experience. It's broken in 0.8.10 and latest.

I just tested tag 0.8.10, SSH server works out of box..

**Comment by prologic** [comment]
Did you use a volume (Docker 1.9 feature)?


James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by bkcsoft** [comment]
it's in `/data/gogs/logs/auth.log`

**Comment by prologic** [comment]
No; there is no ``auth.log`` configured for the Docker image:
```#!bash
$ dke f205dcab24d9 /bin/bash
bash-4.3# cd /data/gogs/log/
bash-4.3# ls -l
total 36
-rw-r-----    1 git      git           3991 Jan 14 07:43 gogs.log
-rw-r-----    1 git      git              0 Jan  2 00:19 http.log
lrwxrwxrwx    1 git      git             14 Jan  3 00:24 log -> /data/gogs/log
-rw-r--r--    1 git      git          30474 Jan 14 09:14 xorm.log
```

Unless you run this on stock standard Linux host; sure :)

**Comment by Unknwon** [comment]
Can you check permission for `/data/git/.ssh` and all its files?

**Comment by Unknwon** [comment]
> Did you use a volume (Docker 1.9 feature)?

I'm not pro on Docker... I volumed my directory but don't know what it differs from older version?

**Comment by bkcsoft** [comment]
I'm running gogs in Docker 1.9.1 with a named volume and ssh-keys for access, works just fine. though I noticed now that ssh is indeed not producing any logs :unamused:

**Comment by prologic** [comment]
I had already checked permissions; but here's a paste:
```#!bash
$ dke f205dcab24d9 /bin/bash
bash-4.3# cd /data/gogs/log/
bash-4.3# ls -l
total 36
-rw-r-----    1 git      git           3991 Jan 14 07:43 gogs.log
-rw-r-----    1 git      git              0 Jan  2 00:19 http.log
lrwxrwxrwx    1 git      git             14 Jan  3 00:24 log -> /data/gogs/log
-rw-r--r--    1 git      git          30474 Jan 14 09:14 xorm.log
bash-4.3# cd /data/git/
bash-4.3# ls -la
total 20
drwxr-xr-x    4 git      git           4096 Jan 14 07:43 .
drwxr-xr-x    5 git      git           4096 Jan  2 00:16 ..
-rw-r--r--    1 git      git             71 Jan 14 07:43 .gitconfig
drwx------    2 git      git           4096 Jan 14 07:31 .ssh
lrwxrwxrwx    1 git      git              9 Jan  3 00:24 git -> /data/git
drwxr-xr-x    3 git      git           4096 Jan  2 00:20 gogs-repositories
bash-4.3# cd .ssh
bash-4.3# ls -la
total 16
drwx------    2 git      git           4096 Jan 14 07:31 .
drwxr-xr-x    4 git      git           4096 Jan 14 07:43 ..
-rw-------    1 git      git            546 Jan 14 07:33 authorized_keys
-rw-------    1 git      git             23 Jan  2 00:16 environment
bash-4.3#
```

As @bkcsoft mentions though; in this Docker image no sshd logging seems to have been set up; so without horribly hacking the running container I can't easily debug this :/

**Comment by prologic** [comment]
Note there should be no theoretical difference between a "bind mounted volume" (``docker run -v /path/on/host:/path/on/container``) to "named volumes" (``docker volume create -d foo gogs; docker run -v foo:/data ...``).

**Comment by bojoyzhou** [comment]
Ensure that all keys are begin with `command="/home/git/gogs/gogs` in `authorized_keys`.
I have met the same question because of i have tried to add my `id_rsa.pub` to `authorized_keys` by myself.
It works after i delete all keys except begin with `command="/home/git/gogs/gogs` in `authorized_keys`

**Comment by prologic** [comment]
Okay so I finally figured this out; and to be honest I feel a bit stupid.

I will submit a PR to close this issue to update relevant docs on this.

The TL;DR is that if you run gogs via Docker; *most* Docker hosts already have an sshd service running and listening on the default port 22 for management of the host(s). This is "incompatible" with the default gogs sshd setting of port 22. You have to change this in the initial setup to something like 2222 and run gogs with: ``docker run -d 2222:22 ...`` because this setting doesn't affect the listening port but the SSH URI(s) presented in the UI but it's important to "expose" the right outer port in Docker!

**Comment by Unknwon** [comment]
Oh, OK... that's great news.... :joy:

You can help update [README](https://github.com/gogits/gogs/blob/master/docker/README.md) by adding a notice.

**Comment by ivanmarban** [comment]
Since you are mapping port XX of host machine to port 22 of container, you will need to change the URI(s) presented by de UI on Gogs while using SSH protocol.

You will need to add `git clone ssh://` in front of the command and put port XX in front of your user name.

`git clone git@host:username/repo.git`
becomes 
`git clone ssh://git@host:port/username/repo.git`

**Comment by prologic** [comment]
I'll have a PR ready soon to close this issue so others don't make the same
mistake I did :)


James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by cweagans** [comment]
Note that you can also add an entry to your `~/.ssh/config` file:

```
Host git.example.com
    Port 2222
```

and then use the normal `git clone git@git.example.com:myuser/myproject.git` command.

**Comment by prologic** [comment]
Good idea; probably helpful to make a note of this too :) As most Docker
"machines" usually have a listening SSH port on the default port anyway;
it's easy to forget to bind the container port so something not 22 :)


James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by prologic** [comment]
CLosed by 894946c319aa0e961388227c7cff11b1392edfc7

**Comment by Unknwon** [comment]
:100:
